### PR TITLE
Add action base class

### PR DESCRIPTION
**Pull request by danielamorse**
## Jira

 - http://vjira2:8080/browse/BDS-534

## Summary

Add a base class that contains the shared parts of links and buttons, and extend *that* when adding new "actionable" components, rather than extending the link or button components themselves.

## Details

TBD

## How to test

See [files changed](https://github.com/bolt-design-system/bolt/pull/1000/files). I'm looking for feedback on:
 - method of setting `rootElementTags`
 - should this be in renderers or make it and element called `action`?
 - should I try to refactor the chunks of the `render` method into something reusable (handling slots perhaps)? Not worth it?

### CODE REVIEW

File: packages/components/bolt-button/src/button.standalone.js [hunk 1]
```diff
@@ -1,17 +1,6 @@
...
-  watchForComponentMutations,
-} from '@bolt/core/utils';
-import {
-  withLitHtml,
-  html,
-  render,
-} from '@bolt/core/renderers/renderer-lit-html';
+import { props, define } from '@bolt/core/utils';
+import { html, render } from '@bolt/core/renderers/renderer-lit-html';
+import { BoltAction } from '@bolt/core/renderers/bolt-action';
```
Comment (sghoweri) [inline packages/components/bolt-button/src/button.standalone.js]: I'd recommend moving this into the `@bolt/core/elements` folder -- this isn't a renderer, it's a lower level (opinionated) element that a bunch of components extend. 🙂

File: packages/components/bolt-button/src/button.standalone.js [hunk 2]
```diff
@@ -37,86 +26,17 @@ class BoltButton extends withLitHtml() {
...
-
     onClick: props.string, // Managed by base class
     onClickTarget: props.string, // Managed by base class
   };
 
   // https://github.com/WebReflection/document-register-element#upgrading-the-constructor-context
   constructor(self) {
     self = super(self);
-    self.useShadow = hasNativeShadowDomSupport;
+    self.rootElementTags = ['button', 'a'];
```
Comment (sghoweri) [inline packages/components/bolt-button/src/button.standalone.js]: Add comment about what these `rootElementTags` to -- or at least link to where we explain a little more about these.

File: packages/core/utils/get-component-root-element.js [hunk 3]
```diff
@@ -0,0 +1,36 @@
+import { sanitizeBoltClasses } from './sanitize-classes';
+
+export function getComponentRootElement(nodeList, tags) {
```
Comment (sghoweri) [inline packages/core/utils/get-component-root-element.js]: Docs on this please!!

File: packages/core/renderers/bolt-base.js [hunk 4]
```diff
@@ -53,6 +54,27 @@ export function BoltBase(Base = HTMLElement) {
       }
     }
 
+    validateProps(schema, propData) {
```
Comment (sghoweri) [inline packages/core/renderers/bolt-base.js]: @danielamorse I'm thinking we should just have an internal `schema` prop that we add the external schema data to -- `this.schema`; Very close parallel to what we did previously with the component-specific stylesheet.

File: packages/core/renderers/bolt-action.js [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    disabled: props.boolean,
+    onClick: props.string, // Managed by base class
+    onClickTarget: props.string, // Managed by base class
+  };
+
+  // https://github.com/WebReflection/document-register-element#upgrading-the-constructor-context
+  constructor(self) {
+    self = super(self);
+    self.useShadow = hasNativeShadowDomSupport;
+    self.rootElementTags = [];
```
Comment (sghoweri) [inline packages/core/renderers/bolt-action.js]: @danielamorse we need to make sure we move this line and the `if (this._wasInitiallyRendered === false) {` logic down below into a separate component decorator (vs another Class we have to extend). 

The url, target, disabled, and click "stuff" that makes up BoltAction should be separated out from the root element work (ex. blockquote needs this rootElement logic as well)

File: packages/components/bolt-button/src/button.standalone.js (hunk 1 above)
Comment (mikemai2awesome) [reply packages/components/bolt-button/src/button.standalone.js]: @sghoweri I find it weird that there are elements under core and there are elements under global, and they are vastly different things.

Comment (danielamorse) [reply packages/components/bolt-button/src/button.standalone.js]: @sghoweri moved to `@bolt/core/elements` [here](https://github.com/bolt-design-system/bolt/pull/1000/commits/5ecc51ae92b5d6c45b34a2d842fef406d3c19f93).

File: packages/components/bolt-button/src/button.standalone.js (hunk 2 above)
Comment (danielamorse) [reply packages/components/bolt-button/src/button.standalone.js]: @sghoweri after refactoring, added comment [here](https://github.com/bolt-design-system/bolt/pull/1000/commits/2278f9ab60ed3a8ea9b1453d4b792566400d75c9#diff-ad8093a42b9e5fc5b748d20b6e11b078R14). Further comments in [the helper function file itself](https://github.com/bolt-design-system/bolt/blob/2278f9ab60ed3a8ea9b1453d4b792566400d75c9/packages/core/utils/get-component-root-element.js).

File: packages/core/utils/get-component-root-element.js (hunk 3 above)
Comment (danielamorse) [reply packages/core/utils/get-component-root-element.js]: @sghoweri [Added](https://github.com/bolt-design-system/bolt/blob/2278f9ab60ed3a8ea9b1453d4b792566400d75c9/packages/core/utils/get-component-root-element.js) 😄

File: packages/core/renderers/bolt-base.js (hunk 4 above)
Comment (danielamorse) [reply packages/core/renderers/bolt-base.js]: @sghoweri refactored into internal `schema` prop, [see commit](https://github.com/bolt-design-system/bolt/pull/1000/commits/22230537d97223e2ff50fedc7901d142ac4bdeb1).

File: packages/core/renderers/bolt-action.js (hunk 5 above)
Comment (danielamorse) [reply packages/core/renderers/bolt-action.js]: @sghoweri refactored into Class decorator, [see commit](https://github.com/bolt-design-system/bolt/pull/1000/commits/2278f9ab60ed3a8ea9b1453d4b792566400d75c9).

**Pull request merged**